pkg/ingester/retention.go

```diff
@@ -39,183 +66,363 @@ func defaultRetentionPolicy() retentionPolicy {
 	if err != nil {
-		return err
+		level.Error(dc.logger).Log(
+			"msg", "failed run high disk cleanup, could not read tenant ids",
+			"err", err,
+		)
+		return 0, 0, true
 	}
 
-	for volumeStatsCurrent.HighDiskUtilization && len(blocks) > 0 && ctx.Err() == nil {
-		// When disk utilization is not lower since the last loop, we end the
-		// cleanup there to avoid deleting all blocks when disk usage reporting
-		// is delayed.
-		if volumeStatsPrev != nil && volumeStatsPrev.BytesAvailable >= volumeStatsCurrent.BytesAvailable {
-			level.Warn(e.logger).Log("msg", "disk utilization is not lowered by deletion of a block, pausing until next cycle")
+	blocks := make([]*tenantBlock, 0)
+	for _, tenantID := range tenantIDs {
+		tenantBlocks, err := dc.blockManager.GetBlocksForTenant(ctx, tenantID)
+		if err != nil {
+			level.Error(dc.logger).Log(
+				"msg", "failed to get blocks for tenant",
+				"tenantID", tenantID,
+				"err", err,
+			)
+
+			// Keep trying to read blocks from other tenants.
+			continue
+		}
+
+		blocks = append(blocks, tenantBlocks...)
+	}
+
+	// Sort by uploaded, then age (oldest first).
+	sort.Sort(blocksByUploadAndAge(blocks))
+
+	prevVolumeStats := &diskutil.VolumeStats{}
+	filesDeleted := 0
+	for _, block := range blocks {
+		// Delete a block.
+		err = dc.blockManager.DeleteBlock(ctx, block)
+		switch {
+		case os.IsNotExist(err):
+			level.Warn(dc.logger).Log(
+				"msg", "failed to delete uploaded block, does not exist",
+				"err", err,
+				"path", block.Path,
+			)
+			return filesDeleted, int(volumeStats.BytesAvailable - prevVolumeStats.BytesAvailable), true
+		case err != nil:
+			level.Error(dc.logger).Log(
+				"msg", "failed run high disk cleanup, could not delete block",
+				"path", block.Path,
+				"err", err,
+			)
+			return filesDeleted, int(volumeStats.BytesAvailable - prevVolumeStats.BytesAvailable), true
+		default:
+			filesDeleted++
+		}
+
+		// Recheck volume stats.
+		prevVolumeStats = volumeStats
+		volumeStats, err = dc.volumeChecker.HasHighDiskUtilization(dc.config.DataPath)
+		if err != nil {
+			level.Error(dc.logger).Log(
+				"msg", "failed to check disk utilization",
+				"err", err,
+			)
 			break
 		}
-		// Delete the oldest block.
-		var b *tenantBlock
-		b, blocks = blocks[0], blocks[1:]
-		level.Warn(e.logger).Log("msg", "disk utilization is high, deleting the oldest block", "path", b.path)
-		if err = e.deleteBlock(b); err != nil {
-			return err
+
+		// Check if deletion should stop.
+		if !volumeStats.HighDiskUtilization || prevVolumeStats.BytesAvailable >= volumeStats.BytesAvailable {
+			level.Warn(dc.logger).Log("msg", "disk utilization is not lowered by deletion of a block, pausing until next cycle")
+			break
 		}
-		volumeStatsPrev = volumeStatsCurrent
-		if volumeStatsCurrent, err = e.volumeChecker.HasHighDiskUtilization(e.dbConfig.DataPath); err != nil {
-			return err
+	}
+
+	return filesDeleted, int(volumeStats.BytesAvailable - prevVolumeStats.BytesAvailable), true
```

Comment:
Since we recheck the disk usage (prevVolumeStats gets re-assigned in the loop), is there a chance we return a partial amount for bytesAvailable here?

Reply:
You're right. This calculation will tell us the number of bytes freed since the last block deletion, which is incorrect. In https://github.com/grafana/pyroscope/pull/2816/commits/e61b35a2bd2f8f98c8cba4658506c2603e54175c I updated the calculation to reflect the actual number of bytes freed.